mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxLastUidDAO.java

```diff
@@ -0,0 +1,69 @@
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.postgres.mail.dao;
+
+import static org.apache.james.mailbox.postgres.mail.PostgresMailboxLastUidModule.PostgresMailboxLastUidTable.MAILBOX_ID;
+import static org.apache.james.mailbox.postgres.mail.PostgresMailboxLastUidModule.PostgresMailboxLastUidTable.MAILBOX_LAST_UID;
+import static org.apache.james.mailbox.postgres.mail.PostgresMailboxLastUidModule.PostgresMailboxLastUidTable.TABLE_NAME;
+import static org.jooq.impl.DSL.coalesce;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.postgres.utils.PostgresExecutor;
+import org.apache.james.mailbox.MessageUid;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.postgres.PostgresMailboxId;
+
+import reactor.core.publisher.Mono;
+
+public class PostgresMailboxLastUidDAO {
+    private final PostgresExecutor postgresExecutor;
+
+    @Inject
+    public PostgresMailboxLastUidDAO(PostgresExecutor postgresExecutor) {
+        this.postgresExecutor = postgresExecutor;
+    }
+
+    public Mono<MessageUid> findLastUidByMailboxId(MailboxId mailboxId) {
+        return postgresExecutor.executeRow(dsl -> Mono.from(dsl.select(MAILBOX_LAST_UID)
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(((PostgresMailboxId) mailboxId).asUuid()))))
+            .flatMap(record -> Mono.justOrEmpty(record.get(MAILBOX_LAST_UID)))
+            .map(MessageUid::of);
+    }
+
+    public Mono<MessageUid> incrementAndGetLastUid(MailboxId mailboxId, int count) {
+        return postgresExecutor.executeRow(dsl -> Mono.from(dsl.update(TABLE_NAME)
+                .set(MAILBOX_LAST_UID, coalesce(MAILBOX_LAST_UID, 0L).add(count))
+                .where(MAILBOX_ID.eq(((PostgresMailboxId) mailboxId).asUuid()))
+                .returning(MAILBOX_LAST_UID)))
```

Comment:
There may be a concurrency issue when this query is used by multiple threads. By default we set autocommit = true. However, the isolation level is still read-commited. Therefore, when this query run multiple times concurrently, the returning value could be the same.

Reply:
IMO as long as it is a single SQL query the storage engine guaranties isolation.

Only unicity is capital. Thus if a data race returns a lower value, we would not override the underlying data: safe.

And concurrency tests seems to pass.